File: backend/refs/matlab/_aux_fun/compute_se_kalman.m
function  se_output = compute_se_kalman(theta_opt,constraints,evaluate_kalman,N_draws)

% This script computes the standard error of parameters and states
% (accounting for parameter estimation error) using the OPG to get the
% variance-covariance matrix of the MLE estimator, and then using the
% monte carlo procedure in Hamilton (1986) and also used by HLW to compute
% the SEs for the states.

% INPUTS

% theta_opt N x 1 vector of the MLE estimate
% constraints N x 1 vector of logical values. If an element is = 1 it means
% the parameter hit the constraint and thus we don't report SEs for it.
% evaluate_kalman: function that computes the Kalman Filter. We are
% interested in the log-likelihood vector. This should be only a function
% of theta.

% First, compute the variance covariance matrix using OPG.
UB = constraints.UB;
LB = constraints.LB;
try
    const_f = constraints.const_f;
catch
    const_f = @(x) -10;
end

cons_vec = (theta_opt==LB)+(theta_opt==UB);

unconst_cols = find(cons_vec==0);
const_cols = find(cons_vec==1);
N_param = length(theta_opt);
out_base = evaluate_kalman(theta_opt);
[T,N_states] = size(out_base.state_contemporaneous); 

%variance covariance matrix ONLY for the unconstrained parameters.
V_mat = compute_opg(theta_opt,constraints,evaluate_kalman);

C_mat_aux = chol(V_mat,'lower'); %this satisfies C_mat*C_mat*; 
C_mat = zeros(N_param,N_param);
C_mat(unconst_cols,unconst_cols) = C_mat_aux;
se_vec = NaN(N_param,1);
se_vec(unconst_cols) = sqrt(diag(V_mat));
t_stat_vec =  NaN(N_param,1);
t_stat_vec(unconst_cols) = theta_opt(unconst_cols)./se_vec(unconst_cols);

% compute SEs for the states using the estimator in Hamilton 1986.

% generate parameter draws
theta_draws = theta_opt + C_mat*randn(N_param,N_draws);

% if sum(cons_vec)>0
%     theta_draws = NaN(N_param,N_draws);
%     theta_draws(unconst_cols,:) = theta_draws_aux;
%     theta_draws(const_cols,:) = theta_opt(const_cols); %fix this values at the bound
% else
%     theta_draws = theta_draws_aux;
% end

filter_unc_mat = zeros(T,N_states,N_states);
param_unc_mat = zeros(T,N_states,N_states);

ii=1;
used_draws = 0;
while ii<=N_draws
    theta_i = theta_draws(:,ii);
    % first, check if the draw satisfies the constraints
    if  sum([theta_i>UB; theta_i<LB; (const_f(theta_i)>0)])>0 %if not just go to the next draw
        ii = ii+1;
    else % if it satisfies the constraint, solve the optimizer.
        out_i = evaluate_kalman(theta_i);
        % collect the results. 
        % uncertainty coming from the filter. Smoothed is the only one that
        % makes sense since the MLE uses full sample
        filter_unc_mat = filter_unc_mat + out_i.state_smoothed_pred; 
        % squeeze(out_i.state_smoothed_pred(end,:,:))
        
        % difference of the state from baseline
        state_dif = (out_i.state_smoothed - out_base.state_smoothed);
        % squeeze(out_i.state_smoothed_pred(end,:,:))
        for t=1:T
            param_unc_mat(t,:,:) = squeeze(param_unc_mat(t,:,:))+state_dif(t,:)'*state_dif(t,:); 
        end
         % squeeze(param_unc_mat(end,:,:))

        % once this is done, advance iters.
        used_draws = used_draws+1;
        ii = ii+1;
    end
end

filter_unc = filter_unc_mat/used_draws;
param_unc = param_unc_mat/used_draws;
state_unc_mat = out_base.state_smoothed_pred + param_unc;
%state_unc_mat = filter_unc + param_unc;
%algo = squeeze(filter_unc_mat(end,:,:))
%algo = squeeze(param_unc(end,:,:))
%algo_2 = squeeze(out_base.state_smoothed_pred(end,:,:))
%  squeeze(state_unc_mat(end,:,:))
state_ses = NaN(T,N_states);
state_ses_no_param = NaN(T,N_states);
state_ses_contemp = NaN(T,N_states);
for t=1:T
    state_ses(t,:) = sqrt(diag(squeeze(state_unc_mat(t,:,:))))';
    state_ses_no_param(t,:) = sqrt(diag(squeeze(filter_unc(t,:,:))))';
    state_ses_contemp(t,:) = sqrt(diag(squeeze(out_base.state_contemporaneous_pred(t,:,:))))';
end



se_output.V_mat = V_mat;
se_output.se_vec = se_vec;
se_output.t_stat_vec = t_stat_vec;
se_output.filter_unc = filter_unc; 
se_output.param_unc =  param_unc;
se_output.state_unc_mat =state_unc_mat ;
se_output.state_ses =  state_ses;
se_output.state_ses_no_param =  state_ses_no_param;
se_output.state_ses_contemp =  state_ses_contemp;
end

File: backend/pipeline/_aux_fun/compute_se_kalman.py
from __future__ import annotations

from typing import Any, Callable

import numpy as np

from _aux_fun.compute_opg import compute_opg


def _get(obj: Any, name: str) -> Any:
    if isinstance(obj, dict):
        return obj[name]
    return getattr(obj, name)


def compute_se_kalman(
    theta_opt: np.ndarray,
    constraints,
    evaluate_kalman: Callable[[np.ndarray], Any],
    N_draws: int,
) -> dict[str, np.ndarray]:
    """
    Compute parameter and state standard errors.

    The procedure follows the MATLAB code:
    - use the OPG to compute the covariance matrix of the MLE
    - draw parameters around theta_opt
    - rerun the Kalman filter/smoother at each accepted draw
    - combine filter uncertainty and parameter uncertainty for the states
    """

    theta_opt = np.asarray(theta_opt, dtype=float).reshape(-1)

    UB = np.asarray(_get(constraints, "UB"), dtype=float).reshape(-1)
    LB = np.asarray(_get(constraints, "LB"), dtype=float).reshape(-1)

    try:
        const_f = _get(constraints, "const_f")
    except Exception:
        const_f = lambda x: np.array([-10.0], dtype=float)

    cons_vec = (theta_opt == LB).astype(int) + (theta_opt == UB).astype(int)

    unconst_cols = np.where(cons_vec == 0)[0]
    const_cols = np.where(cons_vec == 1)[0]
    N_param = len(theta_opt)

    out_base = evaluate_kalman(theta_opt)
    state_base = np.asarray(_get(out_base, "state_contemporaneous"), dtype=float)
    T, N_states = state_base.shape

    # Variance covariance matrix ONLY for the unconstrained parameters.
    opg_out = compute_opg(theta_opt, constraints, evaluate_kalman)
    if isinstance(opg_out, dict):
        V_mat = np.asarray(opg_out["V_mat"], dtype=float)
    else:
        V_mat = np.asarray(opg_out, dtype=float)

    # MATLAB: C_mat_aux = chol(V_mat,'lower')
    C_mat_aux = np.linalg.cholesky(V_mat)

    C_mat = np.zeros((N_param, N_param), dtype=float)
    C_mat[np.ix_(unconst_cols, unconst_cols)] = C_mat_aux

    se_vec = np.full(N_param, np.nan, dtype=float)
    if len(unconst_cols) > 0:
        se_vec[unconst_cols] = np.sqrt(np.diag(V_mat))

    t_stat_vec = np.full(N_param, np.nan, dtype=float)
    if len(unconst_cols) > 0:
        t_stat_vec[unconst_cols] = theta_opt[unconst_cols] / se_vec[unconst_cols]

    # Parameter draws around theta_opt.
    theta_draws = theta_opt[:, None] + C_mat @ np.random.randn(N_param, N_draws)

    filter_unc_mat = np.zeros((T, N_states, N_states), dtype=float)
    param_unc_mat = np.zeros((T, N_states, N_states), dtype=float)

    ii = 0
    used_draws = 0

    while ii < N_draws:
        theta_i = theta_draws[:, ii]

        # Skip draws that violate the constraints.
        violates = np.any(theta_i > UB) or np.any(theta_i < LB) or np.any(
            np.asarray(const_f(theta_i), dtype=float).reshape(-1) > 0
        )

        if violates:
            ii += 1
            continue

        out_i = evaluate_kalman(theta_i)

        # Smoothed covariance: uncertainty coming from the filter.
        state_smoothed_pred_i = np.asarray(_get(out_i, "state_smoothed_pred"), dtype=float)
        filter_unc_mat = filter_unc_mat + state_smoothed_pred_i

        # Difference of the smoothed states relative to the baseline.
        state_smoothed_i = np.asarray(_get(out_i, "state_smoothed"), dtype=float)
        state_smoothed_base = np.asarray(_get(out_base, "state_smoothed"), dtype=float)
        state_dif = state_smoothed_i - state_smoothed_base

        for t in range(T):
            param_unc_mat[t, :, :] = (
                param_unc_mat[t, :, :] + np.outer(state_dif[t, :], state_dif[t, :])
            )

        used_draws += 1
        ii += 1

    if used_draws == 0:
        raise RuntimeError("No accepted parameter draws in compute_se_kalman.")

    filter_unc = filter_unc_mat / used_draws
    param_unc = param_unc_mat / used_draws

    # Total state uncertainty: baseline smoothed covariance + parameter uncertainty.
    state_unc_mat = np.asarray(_get(out_base, "state_smoothed_pred"), dtype=float) + param_unc

    state_ses = np.full((T, N_states), np.nan, dtype=float)
    state_ses_no_param = np.full((T, N_states), np.nan, dtype=float)
    state_ses_contemp = np.full((T, N_states), np.nan, dtype=float)

    out_base_state_contemp_pred = np.asarray(_get(out_base, "state_contemporaneous_pred"), dtype=float)

    for t in range(T):
        state_ses[t, :] = np.sqrt(np.diag(state_unc_mat[t, :, :]))
        state_ses_no_param[t, :] = np.sqrt(np.diag(filter_unc[t, :, :]))
        state_ses_contemp[t, :] = np.sqrt(np.diag(out_base_state_contemp_pred[t, :, :]))

    return {
        "V_mat": V_mat,
        "se_vec": se_vec,
        "t_stat_vec": t_stat_vec,
        "filter_unc": filter_unc,
        "param_unc": param_unc,
        "state_unc_mat": state_unc_mat,
        "state_ses": state_ses,
        "state_ses_no_param": state_ses_no_param,
        "state_ses_contemp": state_ses_contemp,
        "used_draws": np.array([used_draws], dtype=int),
        "unconst_cols": unconst_cols,
        "const_cols": const_cols,
    }
